src/recorders.cc

```diff
@@ -0,0 +1,206 @@
+#include "recorders.h"
+#include "buffer_manager.h"
+#include "task_manager.h"
+
+#include <regex>
+
+namespace celerity::detail {
+
+std::string get_buffer_name(const buffer_id bid, const buffer_manager* buff_man) {
+	return buff_man != nullptr ? buff_man->get_buffer_info(bid).debug_name : "";
+}
+
+access_list build_access_list(const task& tsk, const buffer_manager* buff_man, const std::optional<subrange<3>> execution_range = {}) {
+	access_list ret;
+	const auto exec_range = execution_range.value_or(subrange<3>{tsk.get_global_offset(), tsk.get_global_size()});
+	const auto& bam = tsk.get_buffer_access_map();
+	for(const auto bid : bam.get_accessed_buffers()) {
+		for(const auto mode : bam.get_access_modes(bid)) {
+			const auto req = bam.get_mode_requirements(bid, mode, tsk.get_dimensions(), exec_range, tsk.get_global_size());
+			ret.push_back({bid, get_buffer_name(bid, buff_man), mode, req});
+		}
+	}
+	return ret;
+}
+
+reduction_list build_reduction_list(const task& tsk, const buffer_manager* buff_man) {
+	reduction_list ret;
+	for(const auto& reduction : tsk.get_reductions()) {
+		ret.push_back({reduction.rid, reduction.bid, get_buffer_name(reduction.bid, buff_man), reduction.init_from_buffer});
+	}
+	return ret;
+}
+
+task_dependency_list build_task_dependency_list(const task& tsk) {
+	task_dependency_list ret;
+	for(const auto& dep : tsk.get_dependencies()) {
+		ret.push_back({dep.node->get_id(), dep.kind, dep.origin});
+	}
+	return ret;
+}
+
+// removes initial template qualifiers to simplify, and escapes '<' and '>' in the given name,
+// so that it can be successfully used in a dot graph label that uses HTML, and is hopefully readable
+std::string simplify_and_escape_name(const std::string& name) {
+	// simplify
+	auto first_opening_pos = name.find('<');
+	auto namespace_qual_end_pos = name.rfind(':', first_opening_pos);
+	auto simplified = namespace_qual_end_pos != std::string::npos ? name.substr(namespace_qual_end_pos + 1) : name;
+	// escape
+	simplified = std::regex_replace(simplified, std::regex("<"), "&lt;");
+	return std::regex_replace(simplified, std::regex(">"), "&gt;");
+}
+
+task_record::task_record(const task& from, const buffer_manager* buff_mngr)
+    : tid(from.get_id()), debug_name(simplify_and_escape_name(from.get_debug_name())), cgid(from.get_collective_group_id()), type(from.get_type()),
+      geometry(from.get_geometry()), reductions(build_reduction_list(from, buff_mngr)), accesses(build_access_list(from, buff_mngr)),
+      side_effect_map(from.get_side_effect_map()), dependencies(build_task_dependency_list(from)) {}
+
+void task_recorder::record_task(const task& tsk) {
+	CELERITY_TRACE("Recording task {}", tsk.get_id());
```

Comment:
Should we really log every time a task / command is being recorded? This effectively doubles the number of messages we produce.

Reply:
Hmm, good point. It's only `TRACE` so I didn't think much about it, but we also currently log more useful stuff in `TRACE`.

Reply:
Resolved by removing the tracing. It's not really valuable outside of developing the recording infrastructure, which should hopefully be simple enough so that it's hard to break in a way which needs such a crude debugging tool.